server/src/repositories/asset.repository.ts

```diff
@@ -117,59 +117,89 @@ interface GetByIdsRelations {
 export class AssetRepository {
   constructor(@InjectKysely() private db: Kysely<DB>) {}
 
-  async upsertExif(exif: Insertable<AssetExifTable>): Promise<void> {
-    const value = { ...exif, assetId: asUuid(exif.assetId) };
-    await this.db
-      .insertInto('asset_exif')
-      .values(value)
-      .onConflict((oc) =>
-        oc.column('assetId').doUpdateSet((eb) =>
-          removeUndefinedKeys(
-            {
-              description: eb.ref('excluded.description'),
-              exifImageWidth: eb.ref('excluded.exifImageWidth'),
-              exifImageHeight: eb.ref('excluded.exifImageHeight'),
-              fileSizeInByte: eb.ref('excluded.fileSizeInByte'),
-              orientation: eb.ref('excluded.orientation'),
-              dateTimeOriginal: eb.ref('excluded.dateTimeOriginal'),
-              modifyDate: eb.ref('excluded.modifyDate'),
-              timeZone: eb.ref('excluded.timeZone'),
-              latitude: eb.ref('excluded.latitude'),
-              longitude: eb.ref('excluded.longitude'),
-              projectionType: eb.ref('excluded.projectionType'),
-              city: eb.ref('excluded.city'),
-              livePhotoCID: eb.ref('excluded.livePhotoCID'),
-              autoStackId: eb.ref('excluded.autoStackId'),
-              state: eb.ref('excluded.state'),
-              country: eb.ref('excluded.country'),
-              make: eb.ref('excluded.make'),
-              model: eb.ref('excluded.model'),
-              lensModel: eb.ref('excluded.lensModel'),
-              fNumber: eb.ref('excluded.fNumber'),
-              focalLength: eb.ref('excluded.focalLength'),
-              iso: eb.ref('excluded.iso'),
-              exposureTime: eb.ref('excluded.exposureTime'),
-              profileDescription: eb.ref('excluded.profileDescription'),
-              colorspace: eb.ref('excluded.colorspace'),
-              bitsPerSample: eb.ref('excluded.bitsPerSample'),
-              rating: eb.ref('excluded.rating'),
-              fps: eb.ref('excluded.fps'),
-            },
-            value,
+  async upsertExif(exif: Insertable<AssetExifTable>, option: 'update-locked' | 'skip-locked'): Promise<void> {
+    await this.db.transaction().execute(async (tx) => {
+      const { lockedProperties } = await tx
+        .selectFrom('asset_exif')
+        .select('asset_exif.lockedProperties')
+        .where('asset_exif.assetId', '=', exif.assetId)
+        .executeTakeFirstOrThrow();
+
+      const value = { ...exif, assetId: asUuid(exif.assetId) };
+      const getConflictRef = <T extends string>(columnName: T) =>
+        `${option === 'skip-locked' && lockedProperties?.includes(columnName as any) ? 'asset_exif.' : 'excluded.'}${columnName}` as const;
```

Comment:
This is also a bit annoying. I think it might make things nicer to just split up the upsert into an insert, and an update. The update could then only update the columns we're interested in (filtering out locked columns if necessary)

Reply:
Why don't you just remove the locked properties from the `value`? If there are locked properties, you know an exif record already exists.